File: app/src/main/java/com/druk/servicebrowser/ui/viewmodel/ServiceDetailViewModel.java
package com.druk.servicebrowser.ui.viewmodel;

import android.app.Application;
import android.net.nsd.NsdManager;
import android.net.nsd.NsdServiceInfo;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.druk.servicebrowser.BonjourApplication;
import com.druk.servicebrowser.BonjourServiceInfo;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.InetAddress;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.LinkedList;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ServiceDetailViewModel extends AndroidViewModel {

    private static final String TAG = "ServiceDetailVM";
    private static final String HTTP_PROTOCOL = "http";
    private static final String HTTPS_PROTOCOL = "https";

    private final NsdManager nsdManager;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private NsdManager.ServiceInfoCallback serviceInfoCallback;

    private final MutableLiveData<BonjourServiceInfo> serviceInfoLiveData = new MutableLiveData<>();
    private final MutableLiveData<URL> httpUrlLiveData = new MutableLiveData<>();

    public ServiceDetailViewModel(@NonNull Application application) {
        super(application);
        nsdManager = BonjourApplication.getNsdManager(application);
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        if (serviceInfoCallback != null) {
            try {
                nsdManager.unregisterServiceInfoCallback(serviceInfoCallback);
            } catch (IllegalArgumentException ignored) {
            }
            serviceInfoCallback = null;
        }
        executor.shutdownNow();
    }

    public LiveData<BonjourServiceInfo> getServiceInfoLiveData() {
        return serviceInfoLiveData;
    }

    public LiveData<URL> getHttpUrlLiveData() {
        return httpUrlLiveData;
    }

    public void resolve(BonjourServiceInfo service) {
        NsdServiceInfo nsdServiceInfo = new NsdServiceInfo();
        nsdServiceInfo.setServiceName(service.getServiceName());
        nsdServiceInfo.setServiceType(service.getRegType());

        serviceInfoCallback = new NsdManager.ServiceInfoCallback() {
            @Override
            public void onServiceInfoCallbackRegistrationFailed(int errorCode) {
                Log.e(TAG, "ServiceInfoCallback registration failed: " + errorCode);
            }

            @Override
            public void onServiceUpdated(@NonNull NsdServiceInfo nsdServiceInfo) {
                BonjourServiceInfo info = BonjourServiceInfo.fromNsdServiceInfo(nsdServiceInfo, false);
                serviceInfoLiveData.postValue(info);
                checkHttpConnection(info);
            }

            @Override
            public void onServiceLost() {
                Log.d(TAG, "Service lost");
            }

            @Override
            public void onServiceInfoCallbackUnregistered() {
                Log.d(TAG, "ServiceInfoCallback unregistered");
            }
        };

        nsdManager.registerServiceInfoCallback(nsdServiceInfo, Executors.newSingleThreadExecutor(), serviceInfoCallback);
    }

    private void checkHttpConnection(BonjourServiceInfo service) {
        executor.execute(() -> {
            LinkedList<URL> urls = new LinkedList<>();
            for (InetAddress inetAddress : service.getInetAddresses()) {
                try {
                    urls.add(new URL(HTTP_PROTOCOL, inetAddress.getHostAddress(), service.getPort(), ""));
                } catch (MalformedURLException ignored) {
                }
                try {
                    urls.add(new URL(HTTPS_PROTOCOL, inetAddress.getHostAddress(), service.getPort(), ""));
                } catch (MalformedURLException ignored) {
                }
            }

            for (URL url : urls) {
                if (checkURL(url)) {
                    httpUrlLiveData.postValue(url);
                    return;
                }
            }
        });
    }

    private boolean checkURL(@NonNull URL url) {
        try {
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("GET");
            connection.setConnectTimeout(3000);
            connection.setReadTimeout(3000);
            connection.connect();
            int responseCode = connection.getResponseCode();
            if (responseCode == HttpURLConnection.HTTP_OK) {
                return true;
            }
        } catch (IOException ignored) {
        }
        return false;
    }
}

File: app/src/main/java/com/druk/servicebrowser/ui/viewmodel/ServiceDetailViewModel.kt
package com.druk.servicebrowser.ui.viewmodel

import android.app.Application
import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import android.os.Build
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.druk.servicebrowser.BonjourApplication
import com.druk.servicebrowser.BonjourServiceInfo
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.IOException
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL
import java.util.concurrent.Executors

class ServiceDetailViewModel(application: Application) : AndroidViewModel(application) {

    private val nsdManager = BonjourApplication.getNsdManager(application)
    private var serviceInfoCallback: Any? = null

    private val _serviceInfo = MutableStateFlow<BonjourServiceInfo?>(null)
    val serviceInfo: StateFlow<BonjourServiceInfo?> = _serviceInfo

    private val _httpUrl = MutableStateFlow<URL?>(null)
    val httpUrl: StateFlow<URL?> = _httpUrl

    override fun onCleared() {
        super.onCleared()
        if (Build.VERSION.SDK_INT >= 34 && serviceInfoCallback != null) {
            try {
                nsdManager.unregisterServiceInfoCallback(
                    serviceInfoCallback as NsdManager.ServiceInfoCallback
                )
            } catch (_: IllegalArgumentException) {
            }
        }
        serviceInfoCallback = null
    }

    fun resolve(service: BonjourServiceInfo) {
        val nsdServiceInfo = NsdServiceInfo().apply {
            serviceName = service.displayName
            serviceType = service.regType
        }

        if (Build.VERSION.SDK_INT >= 34) {
            resolveLive(nsdServiceInfo)
        } else {
            resolveOnce(nsdServiceInfo)
        }
    }

    private fun resolveLive(nsdServiceInfo: NsdServiceInfo) {
        if (Build.VERSION.SDK_INT < 34) return

        val callback = object : NsdManager.ServiceInfoCallback {
            override fun onServiceInfoCallbackRegistrationFailed(errorCode: Int) {
                Log.e(TAG, "ServiceInfoCallback registration failed: $errorCode")
            }

            override fun onServiceUpdated(nsdServiceInfo: NsdServiceInfo) {
                val info = BonjourServiceInfo.fromNsdServiceInfo(nsdServiceInfo, false)
                _serviceInfo.value = info
                checkHttpConnection(info)
            }

            override fun onServiceLost() {
                Log.d(TAG, "Service lost")
            }

            override fun onServiceInfoCallbackUnregistered() {
                Log.d(TAG, "ServiceInfoCallback unregistered")
            }
        }

        serviceInfoCallback = callback
        nsdManager.registerServiceInfoCallback(nsdServiceInfo, Executors.newSingleThreadExecutor(), callback)
    }

    private fun resolveOnce(nsdServiceInfo: NsdServiceInfo) {
        @Suppress("DEPRECATION")
        nsdManager.resolveService(nsdServiceInfo, object : NsdManager.ResolveListener {
            override fun onResolveFailed(serviceInfo: NsdServiceInfo, errorCode: Int) {
                Log.e(TAG, "Resolve failed: $errorCode")
            }

            override fun onServiceResolved(serviceInfo: NsdServiceInfo) {
                val info = BonjourServiceInfo.fromNsdServiceInfo(serviceInfo, false)
                _serviceInfo.value = info
                checkHttpConnection(info)
            }
        })
    }

    private fun checkHttpConnection(service: BonjourServiceInfo) {
        viewModelScope.launch {
            val url = withContext(Dispatchers.IO) {
                val urls = mutableListOf<URL>()
                for (inetAddress in service.inetAddresses) {
                    try {
                        urls.add(URL(HTTP_PROTOCOL, inetAddress.hostAddress, service.port, ""))
                    } catch (_: MalformedURLException) {
                    }
                    try {
                        urls.add(URL(HTTPS_PROTOCOL, inetAddress.hostAddress, service.port, ""))
                    } catch (_: MalformedURLException) {
                    }
                }
                urls.firstOrNull { checkURL(it) }
            }
            if (url != null) {
                _httpUrl.value = url
            }
        }
    }

    private fun checkURL(url: URL): Boolean {
        return try {
            val connection = url.openConnection() as HttpURLConnection
            connection.requestMethod = "GET"
            connection.connectTimeout = 3000
            connection.readTimeout = 3000
            connection.connect()
            connection.responseCode == HttpURLConnection.HTTP_OK
        } catch (_: IOException) {
            false
        }
    }

    companion object {
        private const val TAG = "ServiceDetailVM"
        private const val HTTP_PROTOCOL = "http"
        private const val HTTPS_PROTOCOL = "https"
    }
}
